crates/sui-core/src/authority.rs

```diff
@@ -269,6 +274,12 @@ pub struct AuthorityMetrics {
     // Tracks recent average txn queueing delay between when it is ready for execution
     // until it starts executing.
     pub execution_queueing_latency: LatencyObserver,
+
+    // Tracks the rate of transactions become ready for execution in transaction manager.
+    pub txn_ready_rate_tracker: Arc<Mutex<RateTracker>>,
```

Comment:
My guess is that contention on this mutex won't be an issue right now, because it should be locked very very briefly relative to overall execution time. But I am a bit nervous about adding something that could be an issue later, and which will probably not be easy to notice

Reply:
Currently it is not an issue because there is *almost* no concurrency to it.

Execution driver uses single thread to schedule and spawn new tasks, that's where the execution_rate_tracker is updated, and transaction manager also is entirely single threaded at the moment (every txn requires taking the TransactionManager::Inner lock for scheduling). The reason we need a Mutex here is for reading the rate in the overload_monitor, there we can control it to use relatively long update interval (10s currently) so that the contention for read and writing is rare.

I thought about what we are going to do in the future when execution driver and transaction manager can be multi-threaded. What we can do is to have a rate tracker in each thread to tracking the per-thread rate, and in the overload_monitor, we aggregate them, and maintain a low lock conflict.

I updated the comments to reflect this.

I also tried to only use atomic type to implement RateTracker, and the complexity there is non-trivial.

Please let me know what do you think?

Reply:
I think what I missed before is that it is only locked for reading in the overload monitor task, which runs infrequently, so I agree this is fine